### ISSUE TITLE
[MRESOLVER-614] Collector applies depMgt entries coming from itself onto itself

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-resolver#1287]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** opened **[MRESOLVER-614](https://issues.apache.org/jira/browse/MRESOLVER-614?redirect=false)** and commented

In Maven3 this was not a problem, as Maven3 could add depMgt only on "root level".

In Maven4 with use of new/old transitive Dependency Managers this became a problem. The dependency manager first receives depMgt entries for currently processed node, and then applies it onto same entry, overriding direct dependency versions.

Example with Maven beta-5: it cannot build Resolver master (88a96ac0606d4d2372452a677d9f93c0b55105a4) as bnd-maven-plugin do depend on plexus-utils via path of `bnd-maven-plugin > build-api > plexus-utils` as it explodes due lack of org.plexus.utils.Scanner on classpath. This class was introduced in plexus-utils 1.5.8, and build-api do depend on it. But alas, build-api parent contains depMgt entry for plexus-utils 1.5.5 and is applied onto build-api overriding it's own dependency version.

Basically current resolver code is NOT prepared for any "transitive dependency management" (despite DependencyManagers are there) as node would apply it's own depManagement rules onto itself.

---

**Affects:** 2.0.2

**Remote Links:**
- [GitHub Pull Request #588
](https://github.com/apache/maven-resolver/pull/588)
- [GitHub Pull Request #595
](https://github.com/apache/maven-resolver/pull/595)

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

[didierl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=didierl) and [gnodet](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=gnodet): ping

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Related "experiment" PR in resolver:
https://github.com/apache/maven-resolver/pull/588
PR on Maven to run ITs:
https://github.com/apache/maven/pull/1841

**Comment by jira-importer** [comment]
**[Didier Loiseau](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=didier%20loiseau)** commented

This is the kind of issue I was worried about in MNG-8190…

> But alas, build-api parent contains depMgt entry for plexus-utils 1.5.5 and is applied onto build-api overriding it's own dependency version.

This is really weird, how come the parent overrides the version specified explicitly in the child?

On the other hand, the dependency management is there, so it should be taken into account… As I understand it, your fix is to ignore it when the `<dependencies>` override the version specified in the `<dependencyManagement>`? For me this makes sense indeed, as what we want is to have the same dependencies as when the child was built.

Watch out for optional dependencies too, as I have noticed they behave the same as dependency management.

p.s. it seems you found an old Apache Jira account of mine there 😅

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

No, I disagree with your assessment, and while true that we see "slight" differences between (non-transitive) Maven3 and (transitive) Maven4, I am pretty much sure it is due this issue. In general, transitive depMgt handling should just improve things IMO.
When I say "in parent", I literally meant "in parent POM", but resolver receives model builder built POM (hence, same would be true, if a POM is at n-depth, and contains depMgt entry for Dv1 while it contains dep entry for Dv2... Dv1 would "won". Resolver is not aware that depMgt is "in parent" nor is important. The important bit is that depMgt from node at n-depth should be applied only on n+1 and below depth (to children, so to say, so it's own dependency dependencies, just like in your POM, the depMgt is not overriding your deps).

Anyway, the PR looks good IMO and also seems to work (due this issue, Maven 4 beta-5 is unable to build Maven Resolver master, with this patch it can).

**Comment by jira-importer** [comment]
**[Didier Loiseau](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=didier%20loiseau)** commented

> In general, transitive depMgt handling should just improve things IMO.

Oh I totally agree with that. In fact, I would even argue that, if it causes any issue, the project is probably misconfigured. But some may still consider it a regression. (OTOH, they can continue to use Maven 3)

> Resolver is not aware that depMgt is "in parent" nor is important.

Yes indeed, in this case it is in parent but it could even be in the same pom.

> The important bit is that depMgt from node at n-depth should be applied only on n+1 and below depth (to children, so to say, so it's own dependency dependencies, just like in your POM, the depMgt is not overriding your deps).

That makes sense indeed. (y)

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

So here is a reproducer:
https://github.com/cstamas/MRESOLVER-614

Notice how 3.9.9 claims level3 is 1.0.0, while 4.0.0-beta-5 says is 1.0.1 -- wrongly. As there is this:
https://github.com/cstamas/MRESOLVER-614/blob/main/local-repo/org/apache/maven/it/mresolver614/level2/1.0.0/level2-1.0.0.pom

So what happens is that level2 depMgt overrides its own dep definitions. Basically "in parent" does not matter, just like I assumed.

**Comment by jira-importer** [comment]
**[Tamas Cservenak](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cstamas)** commented

Since then reproducer updated with multiple examples, and explanations. This will make an IT.
